File: tracker/ats/ashby.py
"""Ashby job board posting API.

GET https://api.ashbyhq.com/posting-api/job-board/{slug}?includeCompensation=true

Compensation is nested and optional at every level: a board may send no
`compensation` key, a summary string with no numbers, or tiers whose components
are equity rather than salary. Everything below tolerates all three.
"""
from .. import http
from ..textutil import as_float, as_str, clean, dig, strip_html
from . import FetchResult, NormalizedJob

API = "https://api.ashbyhq.com/posting-api/job-board/{slug}?includeCompensation=true"
SOURCE = "ashby"


def board_url(slug: str) -> str:
    return f"https://jobs.ashbyhq.com/{slug}"


def _components(comp) -> list[dict]:
    """Every salary component anywhere in the compensation blob."""
    if not isinstance(comp, dict):
        return []
    found = list(comp.get("summaryComponents") or [])
    for tier in comp.get("compensationTiers") or []:
        if isinstance(tier, dict):
            found.extend(tier.get("components") or [])
    return [
        c for c in found
        if isinstance(c, dict) and str(c.get("compensationType", "")).lower() == "salary"
    ]


def _pay(comp) -> tuple[float | None, float | None, str | None]:
    lows, highs, currency = [], [], None
    for c in _components(comp):
        low, high = as_float(c.get("minValue")), as_float(c.get("maxValue"))
        if low is not None:
            lows.append(low)
        if high is not None:
            highs.append(high)
        currency = currency or as_str(c.get("currencyCode"))
    return (min(lows) if lows else None, max(highs) if highs else None, currency)


def parse(payload) -> list[NormalizedJob]:
    jobs = []
    for item in (payload or {}).get("jobs") or []:
        if not isinstance(item, dict):
            continue
        external_id = as_str(item.get("id"))
        title = as_str(item.get("title"))
        if not external_id or not title:
            continue
        pay_min, pay_max, currency = _pay(item.get("compensation"))
        body = as_str(item.get("descriptionPlain"))
        remote = item.get("isRemote")
        jobs.append(
            NormalizedJob(
                external_id=external_id,
                title=title,
                url=as_str(item.get("jobUrl")) or "",
                source=SOURCE,
                posted_at=as_str(item.get("publishedAt")),
                location=as_str(item.get("location")) or as_str(dig(item, "address", "postalAddress", "addressLocality")),
                employment_type=as_str(item.get("employmentType")),
                remote=bool(remote) if isinstance(remote, bool) else None,
                jd_text=clean(body) if body else strip_html(item.get("descriptionHtml")),
                pay_min=pay_min,
                pay_max=pay_max,
                pay_currency=currency,
                raw=item,
            )
        )
    return jobs


def fetch(slug: str, *, etag=None, last_modified=None) -> FetchResult:
    resp = http.get(API.format(slug=slug), etag=etag, last_modified=last_modified)
    if resp.not_modified:
        return FetchResult(jobs=None, status=304, not_modified=True,
                           etag=etag, last_modified=last_modified)
    if not resp.ok:
        return FetchResult(jobs=None, status=resp.status,
                           error=resp.error or f"HTTP {resp.status}")
    payload = resp.json()
    if payload is None:
        return FetchResult(jobs=None, status=resp.status, error="unparseable JSON body")
    return FetchResult(jobs=parse(payload), status=resp.status,
                       etag=resp.etag, last_modified=resp.last_modified)

File: internal/ats/ashby.go
package ats

import (
	"fmt"
	"strings"

	"github.com/Vaivaswat2244/job-tracker/internal/httpx"
	"github.com/Vaivaswat2244/job-tracker/internal/textutil"
)

// Ashby reads the job board posting API:
//
//	GET https://api.ashbyhq.com/posting-api/job-board/{slug}?includeCompensation=true
//
// Compensation is nested and optional at every level: a board may send no
// `compensation` key, a summary string with no numbers, or tiers whose
// components are equity rather than salary. Everything below tolerates all three.
type Ashby struct{}

const ashbyAPI = "https://api.ashbyhq.com/posting-api/job-board/%s?includeCompensation=true"

func (Ashby) Name() string { return "ashby" }

func (Ashby) BoardURL(slug string) string { return "https://jobs.ashbyhq.com/" + slug }

// salaryComponents collects every salary component anywhere in the
// compensation blob, discarding equity and other non-salary types.
func salaryComponents(comp any) []map[string]any {
	blob, ok := comp.(map[string]any)
	if !ok {
		return nil
	}

	var found []any
	if summary, ok := blob["summaryComponents"].([]any); ok {
		found = append(found, summary...)
	}
	if tiers, ok := blob["compensationTiers"].([]any); ok {
		for _, raw := range tiers {
			if tier, ok := raw.(map[string]any); ok {
				if components, ok := tier["components"].([]any); ok {
					found = append(found, components...)
				}
			}
		}
	}

	var salary []map[string]any
	for _, raw := range found {
		c, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		kind, _ := textutil.AsString(c["compensationType"])
		if strings.ToLower(kind) == "salary" {
			salary = append(salary, c)
		}
	}
	return salary
}

func pay(comp any) (*float64, *float64, *string) {
	var (
		low, high *float64
		currency  *string
	)
	for _, c := range salaryComponents(comp) {
		if v, ok := textutil.AsFloat(c["minValue"]); ok {
			if low == nil || v < *low {
				low = &v
			}
		}
		if v, ok := textutil.AsFloat(c["maxValue"]); ok {
			if high == nil || v > *high {
				high = &v
			}
		}
		if currency == nil {
			currency = optString(c["currencyCode"])
		}
	}
	return low, high, currency
}

func (a Ashby) Parse(payload any) []NormalizedJob {
	jobs := []NormalizedJob{}
	items, _ := textutil.Dig(payload, "jobs").([]any)
	for _, raw := range items {
		item, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		externalID, hasID := textutil.AsString(item["id"])
		title, hasTitle := textutil.AsString(item["title"])
		if !hasID || !hasTitle {
			continue
		}
		url, _ := textutil.AsString(item["jobUrl"])
		payMin, payMax, currency := pay(item["compensation"])

		location := optString(item["location"])
		if location == nil {
			location = optString(textutil.Dig(item, "address", "postalAddress", "addressLocality"))
		}

		// Only a real JSON boolean counts; an absent or non-bool isRemote stays
		// unknown rather than collapsing to false.
		var remote *bool
		if b, ok := item["isRemote"].(bool); ok {
			remote = &b
		}

		jd := ""
		if body, ok := textutil.AsString(item["descriptionPlain"]); ok {
			jd = textutil.Clean(body)
		} else {
			html, _ := textutil.AsString(item["descriptionHtml"])
			jd = textutil.StripHTML(html)
		}

		jobs = append(jobs, NormalizedJob{
			ExternalID:     externalID,
			Title:          title,
			URL:            url,
			Source:         a.Name(),
			PostedAt:       optString(item["publishedAt"]),
			Location:       location,
			EmploymentType: optString(item["employmentType"]),
			Remote:         remote,
			JDText:         jd,
			PayMin:         payMin,
			PayMax:         payMax,
			PayCurrency:    currency,
			Raw:            item,
		})
	}
	return jobs
}

func (a Ashby) Fetch(slug, etag, lastModified string) FetchResult {
	resp := httpx.Get(fmt.Sprintf(ashbyAPI, slug), httpx.Options{
		ETag: etag, LastModified: lastModified,
	})
	return finish(a, resp, etag, lastModified, func(payload any) (any, string) {
		if payload == nil {
			return nil, "unparseable JSON body"
		}
		return payload, ""
	})
}
